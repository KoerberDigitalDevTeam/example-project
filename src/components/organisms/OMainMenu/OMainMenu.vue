<template>
  <div>
    <ANavigationDrawer
      v-model="navDrawer"
      v-testId.main-menu
      app
      color="dBlack"
      :mini-variant="isCollapsed"
      class="o-main-menu__a-navigation-drawer"
      @update:mini-variant="handleCollapseMenu"
    >
      <div class="o-main-menu__header">
        <div v-testId.main-menu.header class="o-main-menu__header-logo">
          <div v-if="!isCollapsed">
            <img v-testId.main-menu.app-logo :src="logo" alt="logo" class="o-main-menu__logo" />
          </div>
        </div>

        <v-list-item class="o-main-menu__header-user-profile">
          <v-list-item-avatar
            v-testId.main-menu.profile-avatar
            min-width="33"
            width="33"
            height="33"
            class="o-main-menu__header-user-avatar"
          >
            <v-icon color="dWhite" large>mdi-account-circle</v-icon>
          </v-list-item-avatar>

          <v-list-item-title v-testId.main-menu.user-name class="o-main-menu__user">
            {{ userName }}
          </v-list-item-title>

          <AButton
            v-show="!$vuetify.breakpoint.mobile"
            v-testId.main-menu.collapse-btn
            class="o-main-menu__collapse-menu-button"
            icon
            color="dWhite"
            :tertiary="true"
            @click.stop="handleCollapseMenu"
          >
            <template>
              <v-icon>mdi-chevron-left</v-icon>
            </template>
          </AButton>
        </v-list-item>
      </div>

      <v-divider></v-divider>

      <v-list dense class="o-main-menu__list">
        <v-list-item
          v-for="(item, index) in menuItems"
          :key="index"
          v-testId.main-menu.menu-item
          :to="item.path"
          active-class="o-main-menu__item--active"
          link
          class="o-main-menu__item"
        >
          <v-list-item-icon class="m-menu_items-icon">
            <v-icon v-testId.main-menu.menu-item-icon dense>{{ item.icon }}</v-icon>
          </v-list-item-icon>

          <v-list-item-content>
            <v-list-item-title v-testId.main-menu.menu-item-title class="o-main-menu__item-title">
              {{ $t(item.title) }}
            </v-list-item-title>
          </v-list-item-content>
        </v-list-item>
      </v-list>
    </ANavigationDrawer>
  </div>
</template>

<script>
import { mdiAccountCircle, mdiChevronLeft } from '@mdi/js';
import { mapState, mapActions } from 'vuex';
import logo from '@/assets/images/koerber_logo_rgb_white.svg';
import AButton from '@/components/atoms/AButton/AButton.vue';
import ANavigationDrawer from '@/components/atoms/ANavigationDrawer/ANavigationDrawer.vue';

export default {
  name: 'OMainMenu',

  components: {
    AButton,
    ANavigationDrawer,
  },

  props: {
    userName: {
      type: String,
      default: '',
    },
    menuItems: {
      type: Array,
      required: true,
    },
  },

  data() {
    return {
      logo,
      isCollapsed: false,
      mdiAccountCircle,
      mdiChevronLeft,
      navDrawer: false,
    };
  },

  computed: {
    ...mapState({
      isSidebarOpenGlobal: (state) => state.global.isSidebarOpen,
    }),
  },

  watch: {
    isSidebarOpenGlobal(newValue) {
      this.navDrawer = newValue;
    },

    navDrawer(newValue) {
      this.toggleSidebar(newValue);
    },
  },

  methods: {
    ...mapActions({ toggleSidebar: 'global/toggleSidebar' }),
    handleCollapseMenu() {
      this.isCollapsed = !this.isCollapsed;
    },
  },
};
</script>

<style lang="scss" scoped>
::v-deep {
  .theme--light.v-navigation-drawer .v-divider {
    .theme--light.v-divider {
      color: $d-white;
    }
  }

  .v-list--dense .v-list-item .v-list-item__title {
    font-weight: 600;
    font-size: 14px;
    line-height: 24px;
    padding-right: 11px;
  }
}

.o-main-menu__header {
  .o-main-menu__header-logo {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 24px 16px 36px;
    height: 88px;
  }

  .o-main-menu__logo {
    height: 40px;
    width: 68px;
  }

  .o-main-menu__header-user-profile {
    height: 73px;
    padding: 0 11px;

    --color: #{$d-white};
    --icon-size: 33px;

    .o-main-menu__header-user-avatar {
      margin-right: 12px;
    }
  }

  .o-main-menu__user {
    font-weight: 600;
    font-size: 14px;
    line-height: 24px;
    color: $d-white;
    padding-right: 16px;
  }

  .o-main-menu__collapse-menu-button {
    --color: #{$d-white};
    width: 40px; // TODO: Check why the button component don't use this style.
  }
}

.o-main-menu__a-navigation-drawer .o-main-menu__list {
  .o-main-menu__item {
    --color: #{$d-white};

    &.v-list-item--active {
      --color: #{$d-black};
    }

    &:not(.v-list-item--active):not(.v-list-item--disabled) {
      & .v-list-item__title {
        color: $d-white;
      }

      &:hover {
        background: rgba(255, 255, 255, 0.1);
      }
    }

    .m-menu_items-icon {
      width: 56px;
      height: 56px;
      margin: 0;
      display: flex;
      justify-content: center;
    }

    &.o-main-menu__item--active {
      background-color: $d-white;
      color: $d-black;

      &::before {
        opacity: 0;
      }
    }
  }
}

.o-main-menu__list {
  padding: 0;

  .v-list-item {
    padding: 0;
  }
}
</style>
