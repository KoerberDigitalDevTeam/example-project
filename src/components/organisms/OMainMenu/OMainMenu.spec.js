import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import Vuex from 'vuex';
import OMainMenu from '@/components/organisms/OMainMenu/OMainMenu.vue';

function createWrapper(options = {}) {
  const localVue = createLocalVue();

  const global = {
    namespaced: true,
    state: {
      isSidebarOpen: false,
    },
    actions: {
      toggleSidebar: jest.fn(),
    },
  };

  localVue.use(Vuex);

  const store = new Vuex.Store({
    modules: {
      global,
    },
  });

  return shallowMount(OMainMenu, {
    localVue,
    store,
    vuetify: new Vuetify(),
    ...options,
  });
}

describe('OMainMenu', () => {
  const menuItems = [
    { path: '/test', icon: 'test', title: 'test' },
    { path: '/test2', icon: 'test2', title: 'test' },
  ];

  describe('Template:', () => {
    describe('Header', () => {
      it('Should render OMainMenu', () => {
        const wrapper = createWrapper({
          propsData: {
            menuItems,
          },
        });

        expect(wrapper.html()).toMatchSnapshot();
      });

      it('Should have a user name in user profile area', () => {
        const wrapper = createWrapper({
          propsData: {
            userName: 'user',
            menuItems,
          },
        });
        const element = wrapper.find('[data-testid=main-menu-user-name]');
        expect(element.text()).toBe('user');
      });
    });

    describe('Content', () => {
      const wrapper = createWrapper({
        propsData: {
          menuItems,
        },
      });

      it('Should render an menu item for each menu items', () => {
        const elements = wrapper.findAll('[data-testid=main-menu-menu-item]');
        expect(elements).toHaveLength(menuItems.length);
      });

      it('Should render each menu item with path', () => {
        const element = wrapper.find('[data-testid=main-menu-menu-item]');
        expect(element.vm.to).toContain(menuItems[0].path);
      });

      it('Should render each menu item with icon', () => {
        const element = wrapper.find('[data-testid=main-menu-menu-item-icon]');
        expect(element.text()).toBe(menuItems[0].icon);
      });

      it('Should render each menu item with title', () => {
        const element = wrapper.find('[data-testid=main-menu-menu-item-title]');
        expect(element.text()).toContain(menuItems[0].title);
      });
    });
  });

  describe('Props Interface:', () => {
    it('Should have userName prop and must be a string', () => {
      expect(OMainMenu.props).toMatchObject({
        userName: {
          type: String,
          default: '',
        },
      });
    });

    it('Should have menuItems prop and must be a array', () => {
      expect(OMainMenu.props).toMatchObject({
        menuItems: {
          type: Array,
          required: true,
        },
      });
    });
  });

  describe('Watch:', () => {
    it("Should isSidebarOpenGlobal set 'navDrawer' with 'newValue'", () => {
      const localThis = {
        navDrawer: false,
      };

      const payload = true;

      OMainMenu.watch.isSidebarOpenGlobal.call(localThis, payload);

      expect(localThis.navDrawer).toBe(payload);
    });

    it("Should navDrawer send action 'toggleSidebar' with 'newValue'", () => {
      const localThis = {
        toggleSidebar: jest.fn(),
      };

      const payload = true;

      OMainMenu.watch.navDrawer.call(localThis, payload);

      expect(localThis.toggleSidebar).toHaveBeenCalledWith(payload);
    });
  });

  describe('Methods:', () => {
    it('Should return true when the menu is collapsed', async () => {
      const localThis = {
        isCollapsed: false,

        menuItems,
      };

      OMainMenu.methods.handleCollapseMenu.call(localThis);
      expect(localThis.isCollapsed).toBe(true);
    });

    it('Should return false when the menu is expanded', async () => {
      const localThis = {
        isCollapsed: true,
        menuItems,
      };

      OMainMenu.methods.handleCollapseMenu.call(localThis);
      expect(localThis.isCollapsed).toBe(false);
    });
  });
});
