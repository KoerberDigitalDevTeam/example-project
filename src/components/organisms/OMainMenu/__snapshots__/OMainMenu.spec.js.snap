// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OMainMenu Template: Header Should render OMainMenu 1`] = `
<div>
  <anavigationdrawer-stub app="" color="dBlack" class="o-main-menu__a-navigation-drawer" data-testid="main-menu">
    <div class="o-main-menu__header">
      <div class="o-main-menu__header-logo" data-testid="main-menu-header">
        <div><img src="" alt="logo" class="o-main-menu__logo" data-testid="main-menu-app-logo"></div>
      </div>
      <v-list-item-stub activeclass="" tag="div" class="o-main-menu__header-user-profile">
        <v-list-item-avatar-stub height="33" minwidth="33" width="33" size="40" class="o-main-menu__header-user-avatar" data-testid="main-menu-profile-avatar">
          <v-icon-stub color="dWhite" large="">mdi-account-circle</v-icon-stub>
        </v-list-item-avatar-stub>
        <v-list-item-title-stub class="o-main-menu__user" data-testid="main-menu-user-name">

        </v-list-item-title-stub>
        <abutton-stub primary="true" tertiary="true" color="dWhite" icon="" class="o-main-menu__collapse-menu-button" style="display: none;" data-testid="main-menu-collapse-btn">
          <v-icon-stub>mdi-chevron-left</v-icon-stub>
        </abutton-stub>
      </v-list-item-stub>
    </div>
    <v-divider-stub></v-divider-stub>
    <v-list-stub tag="div" dense="true" class="o-main-menu__list">
      <v-list-item-stub activeclass="o-main-menu__item--active" link="true" to="/test" tag="div" class="o-main-menu__item" data-testid="main-menu-menu-item">
        <v-list-item-icon-stub class="m-menu_items-icon">
          <v-icon-stub dense="" data-testid="main-menu-menu-item-icon">test</v-icon-stub>
        </v-list-item-icon-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub class="o-main-menu__item-title" data-testid="main-menu-menu-item-title">
            test
          </v-list-item-title-stub>
        </v-list-item-content-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="o-main-menu__item--active" link="true" to="/test2" tag="div" class="o-main-menu__item" data-testid="main-menu-menu-item">
        <v-list-item-icon-stub class="m-menu_items-icon">
          <v-icon-stub dense="" data-testid="main-menu-menu-item-icon">test2</v-icon-stub>
        </v-list-item-icon-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub class="o-main-menu__item-title" data-testid="main-menu-menu-item-title">
            test
          </v-list-item-title-stub>
        </v-list-item-content-stub>
      </v-list-item-stub>
    </v-list-stub>
  </anavigationdrawer-stub>
</div>
`;
